### PR TITLE
Feature/system notifications

### PR DESCRIPTION
**Pull request by davidjgonzalez**
System Notifications for AEM allows Super-Authors to define Notifications to display to users of the AEM Author system, warning them off system activities.

Supports
* Multiple messages
* Dismissible messages (if not dismissible they show up on every page load); you can dismiss PER page load as a user.
* Styles (colors): green, blue, yellow, red
* Time based messages (only show within a time frame)
* Enable/disable

2015-10-19_at_5 59_pm_1024

2015-10-19_at_6 05_pm

### CODE REVIEW

File: bundle/src/main/java/com/adobe/acs/commons/http/injectors/AbstractHtmlRequestInjector.java [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+
+        if (!(servletRequest instanceof HttpServletRequest)
+                || !(servletResponse instanceof HttpServletResponse)) {
+            return false;
+        }
+
+        final HttpServletRequest request = (HttpServletRequest) servletRequest;
+        final HttpServletResponse response = (HttpServletResponse) servletResponse;
+        
+        if (!StringUtils.equalsIgnoreCase("get", request.getMethod())) {
```
Comment (bstopp) [inline bundle/src/main/java/com/adobe/acs/commons/http/injectors/AbstractHtmlRequestInjector.java]: Should use the constant from HttpServlet

File: bundle/src/main/java/com/adobe/acs/commons/http/injectors/AbstractHtmlRequestInjector.java [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+        filterProps.put("service.ranking", String.valueOf(ranking));
+        filterProps.put("pattern", StringUtils.defaultIfEmpty(pattern, ".*"));
+        filterRegistration = ctx.getBundleContext().registerService(Filter.class.getName(), this, filterProps);
+    }
+
+    protected final void registerAsSlingFilter(ComponentContext ctx, int ranking, String pattern) {
+        Dictionary<String, String> filterProps = new Hashtable<String, String>();
+
+        filterProps.put("service.ranking", String.valueOf(ranking));
+        filterProps.put("sling.filter.scope", "REQUEST");
```
Comment (bstopp) [inline bundle/src/main/java/com/adobe/acs/commons/http/injectors/AbstractHtmlRequestInjector.java]: Again, SlingFilterScope enum.

File: content/src/main/content/jcr_root/apps/acs-commons/components/utilities/system-notifications/notification/clientlibs/notification.js [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+            } else {
+                // Nothing has been dismissed, mark this notification as dismissed
+                uids = uid;
+            }
+
+            setSessionCookie('acs-commons-system-notifications', uids);
+        }
+    });
+    
+    function setSessionCookie(name, value) {
```
Comment (bstopp) [inline content/src/main/content/jcr_root/apps/acs-commons/components/utilities/system-notifications/notification/clientlibs/notification.js]: Need to be careful here, this name may collide with other functions; probably doing the same thing but don't want that chance.

File: content/src/main/content/jcr_root/apps/acs-commons/components/utilities/system-notifications/notification/clientlibs/notification.js [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+        document.cookie = name + '=' + value + '; expires=0; path=/;';
+    }
+    
+    function getCookieValue(name) {
```
Comment (bstopp) [inline content/src/main/content/jcr_root/apps/acs-commons/components/utilities/system-notifications/notification/clientlibs/notification.js]: Same problem as other function.

**Comment by bstopp** [comment]
Two minor things, otherwise looks good to me.

File: content/src/main/content/jcr_root/apps/acs-commons/components/utilities/system-notifications/notification/clientlibs/notification.js [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+
+$(function() { 
+    $.get('/etc/acs-commons/notifications/_jcr_content.list.html', function(data) {
```
Comment (joerghoh) [inline content/src/main/content/jcr_root/apps/acs-commons/components/utilities/system-notifications/notification/clientlibs/notification.js]: Is likely to break when AEM is deployed in a context root != /

File: bundle/src/main/java/com/adobe/acs/commons/wcm/notifications/impl/SystemNotificationsImpl.java [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+    private static final String PATH_NOTIFICATIONS = "/etc/acs-commons/notifications";
+
+    private static final String PN_ON_TIME = "onTime";
+    private static final String PN_OFF_TIME = "offTime";
+    private static final String PN_ENABLED = "enabled";
+
+    private static final String INJECT_TEXT =
+            "<script>"
+                    + "if(window === top) {"
+                    + "   window.jQuery || document.write('<script src=\"/etc/clientlibs/granite/jquery.js\"><\\/script>');"
```
Comment (joerghoh) [inline bundle/src/main/java/com/adobe/acs/commons/wcm/notifications/impl/SystemNotificationsImpl.java]: will break when AEM is deployed with a context root != /

**Comment by davidjgonzalez** [comment]
Good call! Thanks @joergh

**Comment by davidjgonzalez** [comment]
@joerghoh fixed on master

**Comment by davidjgonzalez** [comment]
Documentation added to https://github.com/Adobe-Consulting-Services/adobe-consulting-services.github.io/tree/acs-aem-commons/2.1.0

**Pull request merged**